New color picker module - integrated from github.com/martinchrzan/Col…

martinchrzan opened the pull request:
<!-- Enter a brief description/summary of your PR here. What does it fix/what does it change/how was it tested (even manually, if necessary)? -->
## Summary of the Pull Request

**NEW MODULE** - Color Picker - integrated from [github.com/martinchrzan/ColorPicker](https://github.com/martinchrzan/ColorPicker)

![image](https://user-images.githubusercontent.com/11967522/86533449-b1e15b80-bed1-11ea-974f-2b83c5b6a346.png)

Basic functionality provided:
- Color picker appears when activation shortcut pressed (configurable in the settings)
- Color picker follows the mouse cursor and shows the actual color that is below the cursor
- Scroll up will cause Zoom Window to open for a better color picking precision
- Left mouse click will copy that color into a clipboard in a predefined format (setting)
- Changes cursor when picking a color (can be turned off) 

- Color picker is multimonitor/multi DPI aware. It respect monitors' boundaries and stays always in the view (predefined safe zones in top, bottom, left, right sides of a monitor)

- Added the new section into Settings:

![image](https://user-images.githubusercontent.com/11967522/86801539-69ae6e80-c074-11ea-8866-65ab9471ecc8.png)

- Updated installer to contain required files


<!-- Other than the issue solved, is this relevant to any other issues/existing PRs? --> 
## References
This is the first version of a color picker as discussed in
https://github.com/microsoft/PowerToys/issues/864

<!-- Please review the items on the PR checklist before submitting-->
## PR Checklist
* [x] Applies to #864
* [x] CLA signed. If not, go over [here](https://cla.opensource.microsoft.com/microsoft/PowerToys) and sign the CLA
* [ ] Tests added/passed
* [x] Requires documentation to be updated
* [x] I've discussed this with core contributors already. If not checked, I'm ready to accept this work might be rejected in favor of a different grand plan. Issue number where discussion took place: #xxx

<!-- Provide a more detailed description of the PR, other things fixed or any additional comments/features here -->
## Detailed Description of the Pull Request / Additional comments

<!-- Describe how you validated the behavior. Add automated tests wherever possible, but list manual validation steps taken as well -->
## Validation Steps Performed
Manual testing

martinchrzan committed:
New color picker module - integrated from github.com/martinchrzan/ColorPicker

martinchrzan committed:
Trying to fix build in github

niels9001 commented on src/core/Microsoft.PowerToys.Settings.UI/Views/ShellPage.xaml:
```diff
@@ -94,6 +94,12 @@
 
+                <winui:NavigationViewItem x:Uid="Shell_ColorPicker" helpers:NavHelper.NavigateTo="views:ColorPickerPage">
+                    <winui:NavigationViewItem.Icon>
+                        <PathIcon Data="M 3.3333333 7.7083333 7.75 3.4583333 Z M 15.375 19.541667 19.5 15.458333 Z M 1.1666667 1.1666667 3.3333333 7.7083333 19.125 23.5 23.625 19.166667 7.75 3.4583333 Z"/>
```
I'd replace the PathIcon with:

`<FontIcon Glyph="&#xEF3C;"/>`

This is the official Segoe MDL2 Assets eyedropper / colorpicker icon

martinchrzan committed:
Replaced icon in the settings to use font icon instead of path icon

ryanbodrug-microsoft commented on src/modules/colorPicker/ColorPicker/dllmain.cpp:
```diff
@@ -0,0 +1,188 @@
+        settings.set_description(L"Color picker");
+        //  settings.set_description(GET_RESOURCE_STRING(IDS_LAUNCHER_SETTINGS_DESC));
+
+        settings.set_overview_link(L"https://aka.ms/PowerToysOverview_ColorPicker");
```
@crutkas will you be creating this link?

crutkas replied:
link is created

martinchrzan committed:
Merge branch 'master' into user/martinchrzan/ColorPickerModule

# Conflicts:
#	PowerToys.sln

ryanbodrug-microsoft reviewed (commented):
Sorry for the delay on this.  

This feature looks awesome! There are a couple nit things that I think we should be fix if we want to release it to .20. 

1.  It would be great if you could set the the ColorPicker and ColorPickerUI Projects to a depenency of the runner to support an F5 and run scenario.  eg. 
![image](https://user-images.githubusercontent.com/56318517/87450767-cc8f9080-c5b3-11ea-9bab-20fe2d697c55.png)
2.  [New Bug] - The colorpicker.exe process doesn't quit when we terminate PowerToys.exe
3.  [New Bug] - The cursor gets permanently changed for some windows event after shutting down the process. 
4. [Opinion] - We may want to consider a different default Activation Shortcut as I find Ctrl interfers with copy paste functionality. 

Note:  I just kicked off an official [build ](https://github-private.visualstudio.com/microsoft/_build/results?buildId=7230&view=results )and would like to do a quick smoke test from the msi generated there, but I'm able to build and run from the installer locally. 

I'm happy to approve this if we can commit to fixing the minor bugs I mentioned.  @crutkas alternatively  we could consider feature flags to be able to move forward without introducing bugs we don't have capacity for for .20. 

This is awesome work thank you so much!!

ryanbodrug-microsoft commented on src/modules/colorPicker/ColorPickerUI/Mouse/CursorManager.cs:
```diff
@@ -0,0 +1,73 @@
+            try
+            {
+                Registry.SetValue(CursorsRegistryPath, cursorRegistryName, curFile);
+                Win32Apis.SystemParametersInfo(SPI_SETCURSORS, 0, new IntPtr(0), SPIF_SENDCHANGE);
```
Can this be done with  an API or potentially P/Invoking [SetCursor](https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-setcursor) instead of modifying the registry?

martinchrzan replied:
Unfortunately no. SetCursor cannot change the actual "image" used for the cursor, it can only switch between the ones that Windows knows about, therefore we have change registry first.
For more discussion about this:
https://stackoverflow.com/a/15016483/613299

ryanbodrug-microsoft replied:
@martinchrzan Gotcha. I think we need to be be very careful when terminating the process that we aren't putting peoples machines in a state where the only way to fix it would be to modify the registry.

martinchrzan replied:
Yeah, this whole changing cursor is a little tricky but I used such a cursor which you can still use for a while if process dies unexpectedly.
If this happens, you don't have to change registry - you can go to mouse settings and do any cursor change such as - change cursor size or a set cursor - it will restore back to the original one - Windows is keeping a copy of that setting somewhere so it will not be destroyed.
![image](https://user-images.githubusercontent.com/11967522/87464190-ff1ca600-c612-11ea-981f-a0c0fd56a677.png)

That said, there is no real "save" way of changing cursor. I have added the option in the setting where you can turn off changing the cursor if it gives you any problems or you just don't like it.

martinchrzan committed:
Closing ColorPicker.exe when PowerToys process closed, added color picker project into runner dependencies, restoring cursors on exit, added ManagedCommon as a dependency into installer

martinchrzan commented:
@ryanbodrug-microsoft I have addressed your comments:

1. Added as a dependency
2. ColorPicker.exe now monitors its parent process and closes when PowerToys.exe shuts down
3. Not sure about that one, I tried to restore cursors, could you try and see if you can still replicate it?
4. I don't have any preference, this shortcut was picked randomly, we can set any you like as default. 

Thanks a lot for the review, I know it is quite a lot of code to go through :)

ryanbodrug-microsoft commented:
@martinchrzan Thanks for the updates, and sorry for the delay.  Wrt #3 I'll try to find some repro steps for that, and document a new issue.   It wasn't able to run from the installer, and I tracked it down to attempting to load the wrong architecture for MangedCommon in release.  I've pushed those changes in a user branch at commit [cf2ffca](https://github.com/microsoft/PowerToys/commit/cf2ffcaf0be8297f152b4392a8dd0b6db7664b25).  I also created a PR that has a the above change change in master that fixed the broken installer that was preventing me from testing.  I've used this to kick off an official build but that change is already in master.   I'm just going to do 1 quick smoke test on the official build and then I'll sign of -- trusting you to merge in [cf2ffca](https://github.com/microsoft/PowerToys/commit/cf2ffcaf0be8297f152b4392a8dd0b6db7664b25) before completing the PR.  I didn't want to push it directly to your branch without you approving it.

ryanbodrug-microsoft committed:
User/ryanbod/fix colorpicker release (#5046)

* Changing configuration to x64 instead of AnyCPU.   The previous configuration was preventing the ManagedCommon binary from being loaded in Release.

* Updating MSI Installer with new icons (#4998)

martinchrzan commented:
@ryanbodrug-microsoft Thank you for the investigation! I wasn't aware of it. I merged your changes in and waiting for your approval :)

martinchrzan committed:
Adding missed dll into installer

martinchrzan committed:
Fixed potential exception

martinchrzan committed:
Creating settings.json on the first start when there are none, fixed default keyboard shortcut

ryanbodrug-microsoft reviewed (approved):
I validated the latest changes locally!  Thanks for that.   I also kicked off a new build.  It would be good to double check that before you complete the PR, but I'm very confident it will work now based on the changes you just added.  Thanks again for all the work on this :)

martinchrzan committed:
Added ColorPicker.exe.config into installer

martinchrzan committed:
Start filewatcher after default settings file is created

martinchrzan committed:
Merge branch 'master' into user/martinchrzan/ColorPickerModule

martinchrzan committed:
Merge branch 'master' into user/martinchrzan/ColorPickerModule

martinchrzan committed:
Fixing build
